sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -175,6 +175,70 @@ object GpuCast extends Arm {
       cv.stringReplaceWithBackrefs(rule.search, rule.replace)
     })
   }
+
+  def sanitizeStringToIntegralType(input: ColumnVector, ansiEnabled: Boolean): ColumnVector = {
+    // Convert any strings containing whitespace to null values. The input is assumed to already
+    // have been stripped of leading and trailing whitespace
+    val sanitized = withResource(input.containsRe("\\s")) { hasWhitespace =>
+      withResource(hasWhitespace.any()) { any =>
+        if (any.getBoolean) {
+          if (ansiEnabled) {
+            throw new NumberFormatException(GpuCast.INVALID_INPUT_MESSAGE)
+          } else {
+            withResource(GpuScalar.from(null, DataTypes.StringType)) { nullVal =>
+              hasWhitespace.ifElse(nullVal, input)
+            }
+          }
+        } else {
+          input.incRefCount()
+        }
+      }
+    }
+
+    if (ansiEnabled) {
+      // ansi mode only supports simple integers, so no exponents, decimal places, or Inf
+      val regex = "^[+\\-]?[0-9]+$"
+      withResource(sanitized.matchesRe(regex)) { isNumeric =>
+        withResource(isNumeric.not()) { notNumeric =>
+          withResource(notNumeric.any()) { any =>
```

Comment:
Saw the most recent commit changed from `!all()` to this which is less efficient (due to the need to manifest `not()`).  `not()` shouldn't change whether something's null or not and `any()`, like `all()`, ignores nulls, so I'm curious how this fixed anything.  Also are there issues when `notNumeric` is entirely a null batch?

Reply:
I think I sorta answered my own question here a bit -- one main difference is that when the batch is entirely null, `any()` will do the right thing whereas I suspect `all()` will not.  `all()` does _not_ return true if the batch is empty or entirely null.  That means my suggestion at https://github.com/NVIDIA/spark-rapids/pull/2933#discussion_r669977924 probably has the same issue and needs to be reverted to the original behavior.

Reply:
Right, I found this very confusing and spent a lot of time wondering why my null tests were failing. I am starting to think that I should build some convenience methods for some of these operations that we repeat in a lot of the casting code as well. I will review the other change.